### PR TITLE
expose log code to manifest, add InvalideFileLink WarningCode

### PR DESCRIPTION
**Pull request by superyyrrzz**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/Loggers/LogCodesLogListener.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            if (string.IsNullOrEmpty(item.File) || string.IsNullOrEmpty(item.Code))
+            {
+                return;
+            }
+            Codes.AddOrUpdate(
+                item.File,
+                new HashSet<string>(new[] { item.Code }),
+                (k, v) =>
+                    {
+                        v.Add(item.Code);
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Loggers/LogCodesLogListener.cs]: not thread safety

File: src/Microsoft.DocAsCode.Common/Loggers/LogCodesLogListener.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+
+    public class LogCodesLogListener : ILoggerListener
+    {
+        public ConcurrentDictionary<string, ImmutableHashSet<string>> Codes { get; }
+            = new ConcurrentDictionary<string, ImmutableHashSet<string>>();
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Common/Loggers/LogCodesLogListener.cs]: >string [](start = 39, length = 6)

is it OSPlatformCase related? is the key normalized?

File: src/Microsoft.DocAsCode.Common/Path/FilePathComparer.cs [hunk 3]
```diff
@@ -9,19 +9,38 @@ namespace Microsoft.DocAsCode.Common
     public class FilePathComparer
         : IEqualityComparer<string>
     {
+        private bool _ignoreToFullPath;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/Path/FilePathComparer.cs]: >_ignoreToFullPath [](start = 21, length = 17)

readonly?

File: src/Microsoft.DocAsCode.Common/Path/FilePathComparer.cs [hunk 4]
```diff
@@ -9,19 +9,38 @@ namespace Microsoft.DocAsCode.Common
...
             return _stringComparer.Equals(x.ToNormalizedFullPath(), y.ToNormalizedFullPath());
         }
 
         public int GetHashCode(string obj)
         {
-            string path = obj.ToNormalizedFullPath();
+            string path;
+            if (_ignoreToFullPath)
+            {
+                path = obj.ToNormalizedFullPath();
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/Path/FilePathComparer.cs]: >ToNormalizedFullPath [](start = 27, length = 20)

default to ToNormalizedPath?

**Comment by hellosnow** [comment]
:shipit:

**Comment by vicancy** [comment]
>             return _stringComparer.Equals(x.ToNormalizedFullPath(), y.ToNormalizedFullPath());

also need change?

---
Refers to: src/Microsoft.DocAsCode.Common/Path/FilePathComparer.cs:30 in 589174c. [](commit_id = 589174ce9b349808aa58bc920fa6348750ed6b3c, deletion_comment = True)

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**